src/Flysystem/FlysystemAssetStore.php

```diff
@@ -152,30 +179,177 @@ public function setProtectedFilesystem(Filesystem $filesystem)
+     */
+    public function setPublicResolutionStrategy(FileResolutionStrategy $publicResolutionStrategy)
+    {
+        $this->publicResolutionStrategy = $publicResolutionStrategy;
+    }
+
+    /**
+     * @return FileResolutionStrategy
+     * @throws LogicException
+     */
+    public function getProtectedResolutionStrategy()
+    {
+        if (!$this->protectedResolutionStrategy) {
+            $this->protectedResolutionStrategy = Injector::inst()->get(FileResolutionStrategy::class . '.protected');
+        }
+
+        if (!$this->protectedResolutionStrategy) {
+            throw new LogicException("Filesystem misconfiguration error");
+        }
+        return $this->protectedResolutionStrategy;
+    }
+
+    /**
+     * @param FileResolutionStrategy $protectedResolutionStrategy
+     */
+    public function setProtectedResolutionStrategy(FileResolutionStrategy $protectedResolutionStrategy)
+    {
+        $this->protectedResolutionStrategy = $protectedResolutionStrategy;
+    }
+
     /**
      * Return the store that contains the given fileID
      *
      * @param string $fileID Internal file identifier
+     * @deprecated 1.4.0
      * @return Filesystem
      */
     protected function getFilesystemFor($fileID)
     {
-        if ($this->getPublicFilesystem()->has($fileID)) {
-            return $this->getPublicFilesystem();
+        return $this->applyToFileOnFilesystem(
+            function (ParsedFileID $parsedFileID, Filesystem $fs) {
+                return $fs;
+            },
+            $fileID
+        );
+    }
+
+    /**
+     * Generic method to apply an action to a file regardless of what FileSystem it's on. The action to perform should
+     * be provided as a closure expecting the following signature:
+     * ```
+     * function(ParsedFileID $parsedFileID, FileSystem $fs, FileResolutionStrategy $strategy, $visibility)
+     * ```
+     *
+     * `applyToFileOnFilesystem` will try to following steps and call the closure if they are succesfull:
+     * 1. Look for the file on the public filesystem using the explicit fileID provided.
+     * 2. Look for the file on the protected filesystem using the explicit fileID provided.
+     * 3. Look for the file on the public filesystem using the public resolution strategy.
+     * 4. Look for the file on the protected filesystem using the protected resolution strategy.
+     *
+     * If the closure returns `false`, `applyToFileOnFilesystem` will carry on and try the follow up steps.
+     *
+     * Any other value the closure returns (including `null`) will be returned to the calling function.
+     *
+     * @param callable $closure Action to apply.
+     * @param string|array|ParsedFileID $fileID File identication. Can be a string, a file tuple or a ParsedFileID
+     * @param bool $strictHashCheck
+     * @return mixed
+     */
+    private function applyToFileOnFilesystem(callable $closure, $fileID, $strictHashCheck = true)
```

Comment:
I feel like a lot of places we are using $fileID (and subsequently parsing these) we should really be using the tuple.

I'd not mind if this involves us deprecating protected methods that take $fileID as an argument (no public AssetStore methods take $fileID).

Reply:
The more arguments you support, the behaviour of the code becomes more ambiguous and harder to test. Enforce the most normalised form for better stability. :)

Reply:
Yeah ... my thinking when I started writing this function was that I would just start by normalising $fileID to a ParsedID and that would be it. However, as I threw more problems at it, I realised I needed to do things differently based on whatever I started with a string or a ParsedFileID.

I'll split it into two different method.

Reply:
Done ... there's one method that accepts a `ParsedFiledID `and another that expects a File ID string.

I've removed the logic to handle a file file tuple array because, I'm never actually using it.